### PR TITLE
[Doc] add static_graph_execution_doc

### PR DESCRIPTION
**Pull request by kangguangli**
增加了静态图执行过程代码解读文档

### CODE REVIEW

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md [hunk 1]
```diff
@@ -0,0 +1,1789 @@
...
+                 type=None,
+                 inputs=None,
+                 outputs=None,
+                 attrs=None):
+       #这里主要是从op proto获得op的各项具体参数，填充到OpDesc中
+       #这与VarDesc的初始化不同，VarDesc不保存什么，而是保存到proto里
+```
+总结：总的来说，Program的基本表示为OpDesc和VarDesc，分别代表运算逻辑和参与运算的数据，BlockDesc是基础的组织结构，通过 Block Desc的组织来实现控制流(if、while、recurrent等)，ProgramDesc是整体表示，也是用户使用静态图的接口。他们都是C++对象，被封装到Python侧的类中方便调用。
+
+将main_program打印如下：
```
Comment (luotao1) [inline pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:182]: 两张图片没有显示，可以直接以截图形式上传下。
image

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md [hunk 2]
```diff
@@ -0,0 +1,1789 @@
...
+    minumum_last_live_ops.insert(item);
+    vec_instruction_[item].AddGCCheckVar(i);
+  }
+}
+last_live_ops_[i] = minumum_last_live_ops;
+vec_meta_info[i].var_ref_count_ = last_live_ops_[i].size();
+}
+```
+这里消去的逻辑同shrinkDownStream类似，考虑 a -> b，这样的依赖关系，假定他们都使用了同一个变量，那么只需要在b结束后检测即可，不需要在a、b执行结束都检测变量的引用计数。将下面的表反过来就是Op执行结束后需要检查的变量表了。
+
```
Comment (luotao1) [inline pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:1169]: 下面乱码没有显示 
image

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md [hunk 3]
```diff
@@ -0,0 +1,1789 @@
...
+interpreter::WaitEvent(instr_node, place_); //判断Op的前置要求是否完成，与前面对stream的分析有关，执行cudaEventWait
+
+RunInstruction(instr_node);//运行Op的主体部分
+
+CheckGC(instr_node);// 回收变量
+
+interpreter::RecordEvent(instr_node, place_); //记录表明Op已经完成
+```
+首先分析主体部分
```
Comment (luotao1) [inline pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:1456]: image
这段的代码格式有点乱

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md [hunk 4]
```diff
@@ -0,0 +1,1789 @@
...
+```
+
+#### 多流同步示例
+
+##### 1.同一个stream 的FastGC
+```
+A：Y = f(X)
+B: M = f(Y)
+C: N = f(Y)
+```
```
Comment (luotao1) [inline pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:1589]: image
表格没有正确显示

**Review by luotao1 — COMMENTED** [review]
多处表格、图片和代码格式乱，需要调整下，可以看下预览地址。

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md (hunk 1 above)
Comment (kangguangli) [reply pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:183]: 已修改。目前在[我自己的库](https://github.com/kangguangli/paddle_community/blob/static_graph_execution_doc/pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md)里查看是正常的，但是pr里预览还是这样，应该是GitHub的问题，预计合入后是正常的

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md (hunk 4 above)
Comment (kangguangli) [reply pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:1571]: Done

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md (hunk 3 above)
Comment (kangguangli) [reply pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:1437]: Done

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md (hunk 2 above)
Comment (kangguangli) [reply pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:1160]: Done

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md [hunk 5]
```diff
@@ -0,0 +1,1717 @@
...
+with program_guard(main_program, startup_program):
+    x = paddle.static.data(shape = [16, 16], name = 'x')
+    label = paddle.static.data(shape=[16, 1], name = 'label')
+    out = paddle.nn.Linear(in_features=16, out_features=1)(x) 
+    loss = paddle.nn.MSELoss()(out, label)
+    optimizer = Adam()
+    optimizer.minimize(loss)
+
+#执行过程
+place = paddle.CUDAPlace(0) if paddle.is_compiled_with_cuda() else paddle.CPUPlace()
```
Comment (zhiqiu) [inline pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:28]: complete the example code  and make it runable

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md [hunk 6]
```diff
@@ -0,0 +1,1717 @@
...
+  proto::ProgramDesc desc_;
+  std::string cached_hash_str_;
+  
+  std::vector<std::unique_ptr<BlockDesc>> blocks_;
+}
+```
+
+**proto::ProgramDesc**是Program的protobuf对象，可以理解为一种program的二进制存储，它会被用于
+1. 在python侧和C++侧同步op的定义(get_all_op_protos)
+2. 分布式通信时使用
```
Comment (zhiqiu) [inline pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:76]: what's the meaning?

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md [hunk 7]
```diff
@@ -0,0 +1,1717 @@
...
+
+| 分享内容     | 静态图执行过程                         |
+| ------------ | ------------- |
+| 提交作者     | Lv yongkang(@kangguangli)  |
+| 提交时间     | 2022-12-30    |
+| 版本号       | v1.0                        |
+| 依赖飞桨版本  | Paddle2.4.1            |
+| 文件名       | 20221230_static_graph_execution.md |
+
+
```
[the remark this answers is not in the document]
Comment (kangguangli) [reply pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:11]: Add introduction about dynamic graph and static graph in Paddle, which comes from official document

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md (hunk 5 above)
Comment (kangguangli) [reply pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:50]: Done

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md [hunk 8]
```diff
@@ -0,0 +1,1717 @@
...
+```
+
+### 1.1 什么是Program？
+
+神经网络模型在静态图下的表示方式，一个Program对应一个神经网络模型。
+
+**main_program**中保存了神经网络模型.
+**startup_program** 中保存着模型的可训练参数及其初始化的方式，用于初始化模型权重。
+
+静态图的特点是在执行前就已经获得了网络的全部信息，而program就是保存网络信息的对象。
```
[the remark this answers is not in the document]
Comment (kangguangli) [reply pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md:78]: Done

File: pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md (hunk 6 above)
Comment (kangguangli) [reply pfcc/paddle-code-reading/static_graph_execution/20221230_static_graph_execution.md]: Checked  the source code and the second usage "for distributed communication" seems unfounded，so removed it.

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**